rust/src/pgsql/pgsql.rs

```diff
@@ -0,0 +1,957 @@
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use std;
+use nom;
+use std::ffi::CString;
+use std::mem::transmute;
+use crate::applayer::{self, *};
+use super::parser::{self, PgsqlBEMessage, PgsqlFEMessage};
+use crate::core::{self, ALPROTO_UNKNOWN, AppProto, Flow, IPPROTO_TCP};
+
+static mut ALPROTO_PGSQL: AppProto = ALPROTO_UNKNOWN;
+
+#[repr(u8)]
+#[derive(Copy, Clone, PartialOrd, PartialEq, Debug)]
+pub enum PgsqlTransactionState { // a simplified version of this, stored in State - for startup phase, at least
+    ConnectionStart = 0, // TODO [Doubt] or ConnectionRequest? Maybe we don't even need this
+    SslRequest = 1,
+    SslAccepted = 2,  // TODO [Doubt] Maybe we don't even need this
+    SimpleAuthentication = 3,
+    GssEncryptionRequest = 4,
+    AuthenticationGssApi = 5,
+    AuthenticationSasl = 6,
+    AuthenticationSspi = 7,
+    AuthenticationOk = 8,
+    BackendInitialization = 9,
+    ReadyForQuery = 10,
+    NotificationReceived = 11, // TODO not sure if necessary
+    ErrorReceived = 12,
+    SimpleQueryProtocol = 13, // TODO not sure if necessary
+    PossibleInvalidState = 14, // TODO this may be nonsense
+    Termination = 15, // TODO not sure if necessary
+    Finished = 16,
+}
+
+#[derive(Debug)]
+pub struct PgsqlTransaction {
+    pub tx_id: u64,
+    pub state: PgsqlTransactionState,
+    pub requests: Vec<PgsqlFEMessage>,
+    pub responses: Vec<PgsqlBEMessage>,
+
+    de_state: Option<*mut core::DetectEngineState>,
+    events: *mut core::AppLayerDecoderEvents,
+    tx_data: AppLayerTxData,
+}
+
+impl PgsqlTransaction {
+    pub fn new() -> PgsqlTransaction {
+        PgsqlTransaction {
+            tx_id: 0,
+            state: PgsqlTransactionState::ConnectionStart, // TODO question is this the best initialization value?
+            requests: Vec::<PgsqlFEMessage>::new(),
+            responses: Vec::<PgsqlBEMessage>::new(),
+            de_state: None,
+            events: std::ptr::null_mut(),
+            tx_data: AppLayerTxData::new(),
+        }
+    }
+
+    pub fn free(&mut self) {
+        if self.events != std::ptr::null_mut() {
+            core::sc_app_layer_decoder_events_free_events(&mut self.events);
+        }
+        if let Some(state) = self.de_state {
+            core::sc_detect_engine_state_free(state);
+        }
+    }
+}
+
+impl Drop for PgsqlTransaction {
+    fn drop(&mut self) {
+        self.free();
+    }
+}
+#[derive(Debug, PartialEq)]
+pub enum PgsqlStateProgress {
```

Comment:
this looks quite similar to `PgsqlTransactionState`, is there a need for both to exist?

Reply:
They're indeed similar, and I'm still trying to figure that out, to be honest. First, we thought that progress would be tracked per transaction, then, due to how postgresql protocol transactions work, we thought that maybe tracking it just in state would be enough. Currently, basically only `PgsqlStateProgress` is being used. But I would say I must investigate the logs a bit more, to understand if Suri can interpret the traffic correctly with `StateProgress` only.